php/class-coauthors-endpoint.php

```diff
@@ -0,0 +1,306 @@
+			)
+		);
+
+		register_rest_route(
+			static::NS,
+			static::AUTHORS_ROUTE . static::ENDPOINT_POST_ID_REGEX,
+			array(
+				array(
+					'methods'             => 'GET',
+					'callback'            => array( $this, 'get_coauthors' ),
+					'permission_callback' => array( $this, 'can_edit_posts' ),
+					'args'                => array(
+						'post_id' => array(
+							'required'          => true,
+							'type'              => 'number',
+							'validate_callback' => array( $this, 'validate_numeric' ),
+						),
+					),
+				),
+			)
+		);
+
+		register_rest_route(
+			static::NS,
+			static::AUTHORS_ROUTE . static::ENDPOINT_POST_ID_REGEX,
+			array(
+				array(
+					'methods'             => 'POST',
+					'callback'            => array( $this, 'update_coauthors' ),
+					'permission_callback' => array( $this, 'can_edit_coauthors' ),
+					'args'                => array(
+						'post_id'     => array(
+							'required'          => true,
+							'type'              => 'number',
+							'validate_callback' => array( $this, 'validate_numeric' ),
+						),
+						'new_authors' => array(
+							'description' => __( 'Names of coauthors to save.' ),
+							'type'        => 'string',
+							'required'    => false,
+						),
+					),
+				),
+			)
+		);
+	}
+
+	/**
+	 * Search and return authors based on a text query.
+	 *
+	 * @param WP_REST_Request   $request Request object.
+	 * @return WP_REST_Response
+	 */
+	public function get_coauthors_search_results( $request ) {
+		$response = array();
+
+		$search  = strtolower( $request->get_param( 'q' ) );
+		$ignore  = explode( ',', $request->get_param( 'existing_authors' ) );
+		$authors = $this->coauthors->search_authors( $search, $ignore );
+
+		if ( ! empty( $authors ) ) {
+			foreach ( $authors as $author ) {
+				$response[] = $this->_format_author_data( $author );
+			}
+		}
+
+		return rest_ensure_response( $response );
+	}
+
+	/**
+	 * Return a single author.
+	 *
+	 * @param WP_REST_Request   $request Request object.
+	 * @return WP_REST_Response
+	 */
+	public function get_coauthors( $request ) {
+		$response = array();
+
+		$this->_build_authors_response( $response, $request );
+
+		return rest_ensure_response( $response );
+	}
+
+	/**
+	 * Update coauthors.
+	 *
+	 * @param WP_REST_Request   $request Request object.
+	 * @return WP_REST_Response
+	 */
+	public function update_coauthors( $request ) {
+
+		$response = array();
+
+		if ( ! empty( $request->get_param( 'new_authors' ) ) ) {
+			$coauthors = explode( ',', $request->get_param( 'new_authors' ) );
+
+			// Replace all existing authors
+			$this->coauthors->add_coauthors( $request->get_param( 'post_id' ), $coauthors );
+
+			$this->_build_authors_response( $response, $request );
+		}
+
+		return rest_ensure_response( $response );
+	}
+
+	/**
+	 * Validate input arguments.
+	 *
+	 * @param mixed $param Value to validate.
+	 * @return bool
+	 */
+	public function validate_numeric( $param ) {
+		return is_numeric( $param );
```

Comment:
Do we need a wrapper for `is_numeric()` or can we directly just use that as the validation?

Reply:
The wrapper is required because WP passes more arguments to the validation callback than `is_numeric()` accepts, which produces a warning starting in one of the 7.x releases.

Reply:
Ah, okay. Thanks for that, @ethitter.